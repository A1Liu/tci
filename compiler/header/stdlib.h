#ifndef __TCI_STDLIB_H
#define __TCI_STDLIB_H

#define NULL ((void *)0)

typedef unsigned long size_t;

void *malloc(size_t size);
void *realloc(void *buffer, size_t new_size);
void free(void *value);

void exit(int status);

double atof(const char *str);

int atoi(const char *str);
long atol(const char *str);
long long atoll(const char *str);

long strtol(const char *restrict str, char **restrict str_end, int base);
long long strtoll(const char *restrict str, char **restrict str_end, int base);

unsigned long strtoul(const char *restrict str, char **restrict str_end,
                      int base);
unsigned long long strtoull(const char *restrict str, char **restrict str_end,
                            int base);

float strtof(const char *restrict str, char **restrict str_end);
double strtod(const char *restrict str, char **restrict str_end);
// long double strtold(const char *restrict str, char **restrict str_end);

#endif
