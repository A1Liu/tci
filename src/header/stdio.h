#ifndef __TCI_STDIO_H
#define __TCI_STDIO_H

#include <stdarg.h>

typedef unsigned long size_t;
typedef unsigned int fpos_t;
// typedef struct {
//   char state[4];
//   unsigned int idx;
// } fpos_t;

// TODO remove definition from header
typedef struct __tci_file {
  // reentrant lock (required as of C11)
  unsigned long lock;

  // the buffer
  char *buffer;
  unsigned int buffer_pos;
  unsigned int buffer_readable_pos;
  unsigned int buffer_capacity;

  // the current stream position and multibyte conversion state (an object of
  // type mbstate_t)
  fpos_t position;

  // platform-specific identifier of the associated I/O device, such as a file
  // descriptor
  unsigned int fd;

  // error indicator
  int error;

  // Flags
  unsigned short flags;
} FILE;

extern FILE *__tci_stdout;
extern FILE *__tci_stderr;
extern FILE *__tci_stdin;

#define stdout __tci_stdout
#define stderr __tci_stderr
#define stdin __tci_stdin
#define BUFSIZ 1024
#define EOF (-1)

FILE *fopen(const char *filename, const char *mode);
int fclose(FILE *fp);
int remove(const char *filename);

int fputc(int c, FILE *fp);
int fputs(const char *s, FILE *fp);
int fflush(FILE *fp);

int fgetc(FILE *fp);
int ungetc(int c, FILE *fp);
char *fgets(char *restrict str, int count, FILE *restrict stream);
int feof(FILE *fp);

size_t fread(void *ptr, size_t size_of_elements, size_t number_of_elements,
             FILE *a_file);
size_t fwrite(const void *ptr, size_t size_of_elements,
              size_t number_of_elements, FILE *a_file);

void perror(const char *s);

int printf(const char *format, ...);
int fprintf(FILE *stream, const char *format, ...);
int vfprintf(FILE *stream, const char *format, va_list arg);

int sprintf(char *buffer, const char *format, ...);
int snprintf(char *buffer, size_t count, const char *format, ...);

int vprintf(const char *format, va_list va);
int vsnprintf(char *buffer, size_t count, const char *format, va_list va);

int sscanf(const char *restrict buffer, const char *restrict format, ...);

int vsscanf(const char *restrict buffer, const char *restrict format,
            va_list vlist);

int scanf(const char *fmt, ...);
int fscanf(FILE *fp, const char *fmt, ...);
int vfscanf(FILE *fp, const char *fmt, va_list ap);

#endif
