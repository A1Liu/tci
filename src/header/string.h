#ifndef __TCI_STRING_H
#define __TCI_STRING_H

typedef unsigned long size_t;

// Searches for the first occurrence of the character c (an unsigned char) in
// the first n bytes of the string pointed to, by the argument str.
void *memchr(const void *str, int c, size_t n);

// Compares the first n bytes of str1 and str2.
int memcmp(const void *str1, const void *str2, size_t n);

// Copies n characters from src to dest.
void *memcpy(void *dest, const void *src, size_t n);

// Another function to copy n characters from str2 to str1.
void *memmove(void *dest, const void *src, size_t n);

// Copies the character c (an unsigned char) to the first n characters of the
// string pointed to, by the argument str.
void *memset(void *str, int c, size_t n);

// Appends the string pointed to, by src to the end of the string pointed to by
// dest.
char *strcat(char *dest, const char *src);

// Appends the string pointed to, by src to the end of the string pointed to, by
// dest up to n characters long.
char *strncat(char *dest, const char *src, size_t n);

// Searches for the first occurrence of the character c (an unsigned char) in
// the string pointed to, by the argument str.
char *strchr(const char *str, int c);

// Compares the string pointed to, by str1 to the string pointed to by str2.
int strcmp(const char *str1, const char *str2);

// Compares at most the first n bytes of str1 and str2.
int strncmp(const char *str1, const char *str2, size_t n);

// Compares string str1 to str2. The result is dependent on the LC_COLLATE
// setting of the location.
int strcoll(const char *str1, const char *str2);

// Copies the string pointed to, by src to dest.
char *strcpy(char *dest, const char *src);

// Copies up to n characters from the string pointed to, by src to dest.
char *strncpy(char *dest, const char *src, size_t n);

// Calculates the length of the initial segment of str1 which consists entirely
// of characters not in str2.
size_t strcspn(const char *str1, const char *str2);

// Searches an internal array for the error number errnum and returns a pointer
// to an error message string.
char *strerror(int errnum);

// Computes the length of the string str up to but not including the terminating
// null character.
size_t strlen(const char *str);

// Computes the length of the string str up to but not including the terminating
// null character.
size_t strnlen(const char *str, size_t max_length);

// Finds the first character in the string str1 that matches any character
// specified in str2.
char *strpbrk(const char *str1, const char *str2);

// Searches for the last occurrence of the character c (an unsigned char) in the
// string pointed to by the argument str.
char *strrchr(const char *str, int c);

// Calculates the length of the initial segment of str1 which consists entirely
// of characters in str2.
size_t strspn(const char *str1, const char *str2);

// Finds the first occurrence of the entire string needle (not including the
// terminating null character) which appears in the string haystack.
char *strstr(const char *haystack, const char *needle);

// Breaks string str into a series of tokens separated by delim.
char *strtok(char *str, const char *delim);

// Transforms the first n characters of the string src into current locale and
// places them in the string dest.
size_t strxfrm(char *dest, const char *src, size_t n);

#endif
