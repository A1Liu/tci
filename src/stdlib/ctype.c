#include <ctype.h>

int isspace(int c) { return c == ' ' || c == '\t' || c == '\n' || c == '\12'; }

int isdigit(int c) { return c >= '0' && c <= '9'; }

int isalpha(int c) { return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z'); }

int isupper(int c) { return (c >= 'A' && c <= 'Z'); }
