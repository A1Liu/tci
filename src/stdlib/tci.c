#include <stdarg.h>
#include <stddef.h>
#include <stdio.h>
#include <tci.h>

static char *tci_errno_strs[] = {
    /* 0 */ "no error right now!",
    /* 1 */ "file doesn't exist",
    /* 2 */ "file name isn't UTF-8",
    /* 3 */ "too many files (maximum is 4000)",
    /* 4 */ "files too large (maximum total size is 64MB)",
    /* 5 */ "file seeking was out of range",
    /* 6 */ "reading from a terminal output (standard output)",
    /* 7 */ "reading from a terminal output (standard error)",
    /* 8 */ "reading from a terminal output (standard log)",
    /* 9 */ "writing to a terminal input (standard input)",
};

static int tci_ecall_param_counts[] = {
    1, /* TCI_ECALL_EXIT */
    0, /* TCI_ECALL_ARGC */
    0, /* TCI_ECALL_ARGV */

    2, /* TCI_ECALL_OPEN_FD */
    4, /* TCI_ECALL_READ_FD */
    4, /* TCI_ECALL_WRITE_FD */
    3, /* TCI_ECALL_APPEND_FD */
    1  /* TCI_ECALL_FD_LEN */
};

size_t tci_var_size(void *var) {
  __tci_builtin_push(var);
  __tci_builtin_push(var);
  char *begin = __tci_builtin_op("AllocBegin", sizeof(char *));
  char *end = __tci_builtin_op("AllocEnd", sizeof(char *));
  if (begin == NULL || end == NULL) {
    return -1;
  }

  return end - begin;
}

void tci_throw_error(const char *name, const char *message,
                     unsigned int skip_frames) {
  __tci_builtin_push(name);
  __tci_builtin_push(message);
  __tci_builtin_push(skip_frames + 1);
  __tci_builtin_op("Throw", sizeof(void));
}

void *tci_ecall(int ecall_num, ...) {
  if (ecall_num >= sizeof(tci_ecall_param_counts) / sizeof(int)) {
    tci_throw_error(
        "InvalidEcall",
        "Called an invalid environment call (this is an error in TCI)", 1);
  }

  va_list list;
  va_start(list, ecall_num);

  int param_count = tci_ecall_param_counts[ecall_num];
  void *next;
  unsigned int size;

  for (int i = 0; i < param_count; i++) {
    next = __builtin_va_arg(&list);
    size = tci_var_size(next);

    __tci_builtin_push(size);
    __tci_builtin_push(next);
    __tci_builtin_op("PushDyn", sizeof(void));
  }

  va_end(list);

  __tci_builtin_push(ecall_num);
  return __tci_builtin_op("Ecall", sizeof(void *));
}

void tci_perror(char *prefix, int error) {
  if (error < sizeof(tci_errno_strs) / sizeof(char *)) {
    fprintf(stderr, "%s: %s\n", prefix, tci_errno_strs[error]);
  } else {
    fprintf(stderr, "%s: invalid error code (errno = %d)\n", prefix, error);
  }
}

char *tci_strerror(int error) {
  if (error < sizeof(tci_errno_strs) / sizeof(char *)) {
    return tci_errno_strs[error];
  }

  return NULL;
}

void *__builtin_va_arg(va_list *list) {
  __tci_builtin_push(list->__tci_va_current);

  list->__tci_va_current -= 1;

  void *ptr = __tci_builtin_op("TranslateStackId", sizeof(void *));

  return ptr;
}

void __builtin_va_end(va_list *list) { list->__tci_va_current = -1; }
