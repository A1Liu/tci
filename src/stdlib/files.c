#include <errno.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <tci.h>

#define FLAG_NARROW ((uint16_t)0x1)
#define FLAG_WIDE ((uint16_t)0x2)
#define FLAG_LINE_BUF ((uint16_t)0x4)
#define FLAG_FULL_BUF ((uint16_t)0x8)
#define FLAGS_BUF ((uint16_t)0xC)

#define FLAG_IO_INPUT ((uint16_t)0x10)
#define FLAG_IO_OUTPUT ((uint16_t)0x20)
#define FLAG_IO_APPEND ((uint16_t)0x40)
#define FLAGS_IO ((uint16_t)0x70)

#define FLAG_BINARY ((uint16_t)0x80)
#define FLAG_EOF ((uint16_t)0x100)
#define FLAG_CLOSED ((uint16_t)0x200)

#define FLAG_CREATE ((uint16_t)0x400)
#define FLAG_CLEAR ((uint16_t)0x800)

#define FLAG_STDIN_INIT ((uint16_t)0x14)  // FLAG_IO_INPUT + FLAG_LINE_BUF
#define FLAG_STDOUT_INIT ((uint16_t)0x44) // FLAG_IO_APPEND + FLAG_LINE_BUF

#define tci_buffer(fp)                                                         \
  ((char *)((fp)->buffer_capacity <= 8U ? (void *)&(fp)->buffer : (fp)->buffer))

static char __tci_stdin_buffer[BUFSIZ];
static FILE __tci_stdin_struct = {0, __tci_stdin_buffer, 0, 0, BUFSIZ, 0, 0,
                                  0, FLAG_STDIN_INIT};

static char __tci_stdout_buffer[BUFSIZ];
static FILE __tci_stdout_struct = {0, __tci_stdout_buffer, 0, 0, BUFSIZ, 0, 1,
                                   0, FLAG_STDOUT_INIT};

static char __tci_stderr_buffer[BUFSIZ];
static FILE __tci_stderr_struct = {0, __tci_stderr_buffer, 0, 0, BUFSIZ, 0, 2,
                                   0, FLAG_STDOUT_INIT};

FILE *__tci_stdin = &__tci_stdin_struct;
FILE *__tci_stdout = &__tci_stdout_struct;
FILE *__tci_stderr = &__tci_stderr_struct;

static inline uint16_t fopen_mode(const char *mode) {
  if (!strcmp(mode, "r"))
    return FLAG_IO_INPUT;
  if (!strcmp(mode, "rb"))
    return FLAG_IO_INPUT | FLAG_BINARY;
  if (!strcmp(mode, "r+"))
    return FLAG_IO_INPUT | FLAG_IO_OUTPUT;
  if (!strcmp(mode, "rb+") && !strcmp(mode, "rb+"))
    return FLAG_IO_INPUT | FLAG_IO_OUTPUT | FLAG_BINARY;

  if (!strcmp(mode, "w"))
    return FLAG_IO_OUTPUT | FLAG_CREATE | FLAG_CLEAR;
  if (!strcmp(mode, "wb"))
    return FLAG_IO_OUTPUT | FLAG_BINARY | FLAG_CREATE | FLAG_CLEAR;
  if (!strcmp(mode, "w+"))
    return FLAG_IO_INPUT | FLAG_IO_OUTPUT | FLAG_CREATE | FLAG_CLEAR;
  if (!strcmp(mode, "wb+") || !strcmp(mode, "w+b"))
    return FLAG_IO_INPUT | FLAG_IO_OUTPUT | FLAG_BINARY | FLAG_CREATE |
           FLAG_CLEAR;

  if (!strcmp(mode, "a"))
    return FLAG_IO_APPEND | FLAG_CREATE;
  if (!strcmp(mode, "ab"))
    return FLAG_IO_APPEND | FLAG_BINARY | FLAG_CREATE;
  if (!strcmp(mode, "a+"))
    return FLAG_IO_INPUT | FLAG_IO_APPEND | FLAG_CREATE;
  if (!strcmp(mode, "ab+") || !strcmp(mode, "a+b"))
    return FLAG_IO_INPUT | FLAG_IO_APPEND | FLAG_BINARY | FLAG_CREATE;

  tci_throw_error("InvalidFileMode", "file mode was invalid", 2);
}

FILE *fopen(const char *name, const char *mode) {
  uint16_t flags = fopen_mode(mode);
  bool should_create = (flags & FLAG_CREATE) != 0,
       should_clear = (flags & FLAG_CLEAR) != 0;
  flags &= ~(FLAG_CLEAR | FLAG_CREATE); // clear the flags
  flags |= FLAG_FULL_BUF;
  uint32_t open_mode = should_create + should_clear;

  uint64_t fd = tci_ecall(TCI_ECALL_OPEN_FD, name, open_mode);
  switch (fd >> 32) {
  case 0:
    break;

  case TCI_FILE_ERR_DOESNT_EXIST:
    errno = TCI_ERRNO_DOESNT_EXIST;
    return NULL;
  case TCI_FILE_ERR_NAME_NOT_UTF8:
    errno = EILSEQ;
    return NULL;
  case TCI_FILE_ERR_TOO_MANY_FILES:
    errno = TCI_ERRNO_TOO_MANY_FILES;
    return NULL;
  case TCI_FILE_ERR_FILES_TOO_LARGE:
    errno = TCI_ERRNO_FILES_TOO_LARGE;
    return NULL;

  default:
    tci_throw_error("InvalidFileError",
                    "got the wrong file error (this is a bug in TCI)", 0);
  }

  FILE *fp = (uint8_t *)malloc(8U + sizeof(FILE) + BUFSIZ) + 8U;
  fp->buffer = (uint8_t *)(fp + 1);
  fp->buffer_pos = 0;
  fp->buffer_readable_pos = 0;
  fp->buffer_capacity = BUFSIZ;
  fp->position = 0;

  fp->lock = 0;
  fp->fd = fd;
  fp->error = 0;
  fp->flags = flags;

  return fp;
}

int fclose(FILE *fp) {
  const uint16_t output_mask = FLAG_IO_OUTPUT | FLAG_IO_APPEND;
  if (fp->flags & output_mask) {
    int err = fflush(fp);
    if (err)
      return err;
  }

  if (fp->fd < 4) {
    fp->flags &= FLAG_CLOSED;
    return 0;
  }

  free((char *)fp - 8U);
  return 0;
}

// Assumes flags have already been checked for errors
static inline uint64_t tci_write_out(unsigned int fd, fpos_t position,
                                     char *buffer, unsigned int len,
                                     uint16_t flags) {
  if (flags & FLAG_IO_OUTPUT) {
    __tci_builtin_push(fd);
    __tci_builtin_push(position);
    __tci_builtin_push(buffer);
    __tci_builtin_push(len);

    __tci_builtin_push(TCI_ECALL_WRITE_FD);
  } else {
    __tci_builtin_push(fd);
    __tci_builtin_push(buffer);
    __tci_builtin_push(len);

    __tci_builtin_push(TCI_ECALL_APPEND_FD);
  }

  uint64_t result = __tci_builtin_op("Ecall", sizeof(uint64_t));
  switch (result >> 32) {
  case 0:
    break;

  case TCI_FILE_ERR_DOESNT_EXIST:
    tci_throw_error("FileWasDeleted",
                    "file was deleted while we still had it open", 2);

  case TCI_FILE_ERR_OUT_OF_RANGE:
    tci_throw_error("FileIndexInvalid",
                    "wrote outside of file's length (this is a bug in TCI)", 0);

  case TCI_FILE_ERR_FILES_TOO_LARGE:
    return (uint64_t)TCI_ERRNO_FILES_TOO_LARGE << 32;

  default:
    tci_throw_error("InvalidFileError",
                    "got the wrong file error (this is a bug in TCI)", 0);
  }

  return (flags & FLAG_IO_OUTPUT) ? position + len : result;
}

// Assumes flags have already been checked for errors
static inline uint64_t tci_read_in(unsigned int fd, fpos_t position,
                                   char *buffer, unsigned int len,
                                   uint16_t flags) {
  __tci_builtin_push(fd);
  __tci_builtin_push(position);
  __tci_builtin_push(buffer);
  __tci_builtin_push(len);

  __tci_builtin_push(TCI_ECALL_READ_FD);

  uint64_t result = __tci_builtin_op("Ecall", sizeof(uint64_t));
  switch (result >> 32) {
  case 0:
    break;

  case TCI_FILE_ERR_DOESNT_EXIST:
    tci_throw_error("FileWasDeleted",
                    "file was deleted while we still had it open", 2);

  case TCI_FILE_ERR_OUT_OF_RANGE:
    tci_throw_error(
        "FileIndexInvalid",
        "tried to read outside of file's length (this is a bug in TCI)", 0);

  default: {
    tci_perror("internal TCI error", result >> 32);

    tci_throw_error("InvalidFileError",
                    "got the wrong file error (this is a bug in TCI)", 0);
  }
  }

  return result;
}

int fputc(int _c, FILE *fp) {
  if (fp->flags & FLAG_CLOSED)
    tci_throw_error("FileIsClosed", "tried to use an already closed file", 1);

  const uint16_t output_mask = FLAG_IO_OUTPUT | FLAG_IO_APPEND;
  if (!(fp->flags & output_mask))
    tci_throw_error("FileIsNotOutput",
                    "tried to flush a file that isn't an output file", 1);

  char c = _c;
  if (!(fp->flags & FLAGS_BUF)) {
    uint64_t ret = tci_write_out(fp->fd, fp->position++, &c, 1, fp->flags);
    if (ret >> 32) {
      fp->error = ret >> 32;
      return EOF;
    }

    return c;
  }

  if (fp->buffer_pos == fp->buffer_capacity) {
    int ret = fflush(fp);
    if (ret)
      return ret;
  }

  tci_buffer(fp)[fp->buffer_pos++] = c;

  if (c == '\n' && (fp->flags & FLAG_LINE_BUF)) {
    int ret = fflush(fp);
    if (ret)
      return ret;
  }

  return c;
}

int fputs(const char *s, FILE *fp) {
  for (; *s; s++)
    if (fputc(*s, fp) == EOF)
      return EOF;

  return 1;
}

int fflush(FILE *fp) {
  // TODO handle null pointer case

  if (fp->flags & FLAG_CLOSED)
    tci_throw_error("FileIsClosed", "tried to use an already closed file", 1);

  const uint16_t output_mask = FLAG_IO_OUTPUT | FLAG_IO_APPEND;
  if (!(fp->flags & output_mask))
    tci_throw_error("FileIsNotOutput",
                    "tried to flush a file that isn't an output file", 1);

  if (fp->buffer_pos == 0) // nothing to flush
    return 0;

  uint64_t ret = tci_write_out(fp->fd, fp->position, tci_buffer(fp),
                               fp->buffer_pos, fp->flags);
  if (ret >> 32) {
    fp->error = ret >> 32;
    return EOF;
  }

  fp->position = ret;
  fp->buffer_pos = 0;
  fp->buffer_readable_pos = 0;
  return 0;
}

int fgetc(FILE *fp) {
  if (fp->flags & FLAG_CLOSED)
    tci_throw_error("FileIsClosed", "tried to use an already closed file", 1);

  if (!(fp->flags & FLAG_IO_INPUT))
    tci_throw_error("FileIsNotInput",
                    "tried to fgetc a file that isn't an input file", 1);

  if (fp->flags & FLAG_EOF)
    return EOF;

  char *buffer = tci_buffer(fp);

  if (fp->buffer_pos < fp->buffer_readable_pos)
    return buffer[fp->buffer_pos++];

  if (!(fp->flags & FLAGS_BUF)) {
    uint64_t ret = tci_read_in(fp->fd, fp->position++, buffer, 1, fp->flags);

    if (ret >> 32) {
      fp->error = ret >> 32;
      return EOF;
    }

    if (ret == 0) {
      fp->flags |= FLAG_EOF;
      return EOF;
    }

    return *buffer;
  }

  const uint64_t ret =
      tci_read_in(fp->fd, fp->position += fp->buffer_readable_pos, buffer,
                  fp->buffer_capacity, fp->flags);

  if (ret >> 32) {
    fp->error = ret >> 32;
    return EOF;
  }

  if (ret == 0) {
    fp->flags |= FLAG_EOF;
    return EOF;
  }

  fp->buffer_pos = 1;
  fp->buffer_readable_pos = ret;

  return *buffer;
}

int ungetc(int c, FILE *fp) {
  if (fp->flags & FLAG_CLOSED)
    tci_throw_error("FileIsClosed", "tried to use an already closed file", 1);

  if (!(fp->flags & FLAG_IO_INPUT))
    tci_throw_error("FileIsNotInput",
                    "tried to ungetc a file that isn't an input file", 1);

  if (c == EOF)
    return c;

  fp->flags &= ~FLAG_EOF;
  char *buffer = tci_buffer(fp);

  if (fp->buffer_pos)
    return buffer[--fp->buffer_pos] = c;

  return EOF;
}

char *fgets(char *restrict str, int count, FILE *restrict fp) {
  char *original = str;
  for (int i = 1, c; i < count; i++, str++) {
    switch (c = fgetc(fp)) {
    case '\n':
      *str++ = '\n';
    case EOF:
      goto end;
    default:
      *str = c;
    }
  }

end:
  *str = '\0';
  return original;
}

int feof(FILE *fp) { return fp->flags & FLAG_EOF; }

size_t fread(void *ptr, size_t size_of_elements, size_t number_of_elements,
             FILE *a_file);
size_t fwrite(const void *ptr, size_t size_of_elements,
              size_t number_of_elements, FILE *a_file);
