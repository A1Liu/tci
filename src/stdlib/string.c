#include <string.h>
#include <tci.h>

void *memcpy(void *_dest, void *_src, size_t n) {
  char *src = (char *)_src;
  char *dest = (char *)_dest;

  for (int i = 0; i < n; i++)
    dest[i] = src[i];

  return dest;
}

void *memset(void *s, int _c, size_t n) {
  for (unsigned char *p = s, c = (unsigned char)_c; n--;)
    *p++ = c;

  return s;
}

size_t strnlen(const char *str, size_t max_len) {
  for (char *begin = str, *len = max_len; len--; str++)
    if (!*str)
      return str - begin;

  return max_len;
}

size_t strlen(char *str) {
  tci_assert_str(str);

  for (char *begin = str;; str++)
    if (!*str)
      return str - begin;
}

int strcmp(const char *left, const char *right) {
  tci_assert_str(left);
  tci_assert_str(right);

  for (; *left && !(*left - *right); left++, right++)
    ;

  return *left - *right;
}
