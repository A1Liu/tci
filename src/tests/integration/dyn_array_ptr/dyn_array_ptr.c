#include "dyn_array_ptr.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

uint64_t *__dyn_array_capacity_ptr(void *arr) {
  char *capa_loc = ((char *)arr) - sizeof(uint64_t) * 2;
  return (uint64_t *)capa_loc;
}
uint64_t *__dyn_array_len_ptr(void *arr) {
  char *len_loc = ((char *)arr) - sizeof(uint64_t);
  return (uint64_t *)len_loc;
}

void __dyn_array_ensure_add(void *arr_, size_t size) {
  void **arr = (void **)arr_;
  void *buffer = *arr;

  if (buffer == NULL) {
    uint64_t *buffer_begin = malloc(size * 8 + sizeof(uint64_t) * 2);
    buffer = *arr = buffer_begin + 2;
    *__dyn_array_capacity_ptr(buffer) = 8;
    *__dyn_array_len_ptr(buffer) = 0;
    return;
  }

  uint64_t *buffer_begin = __dyn_array_capacity_ptr(buffer);
  uint64_t len = *__dyn_array_len_ptr(buffer);
  uint64_t capacity = *buffer_begin;
  if (len < capacity)
    return;

  capacity = capacity / 2 + capacity;
  buffer_begin = realloc(buffer_begin, size * capacity + sizeof(uint64_t) * 2);
  *buffer_begin = capacity;
  *arr = buffer_begin + 2;
}

uint64_t __dyn_array_add_from(void *arr_, size_t size, void *from, size_t len) {
  void **arr = (void **)arr_;
  char *buffer = *arr;
  uint64_t *buf_begin;
  uint64_t initial_len;

  if (buffer == NULL) {
    initial_len = 0;
    buf_begin = malloc(size * (8 + len) + sizeof(uint64_t) * 2);
    buffer = *arr = buf_begin + 2;
    *__dyn_array_capacity_ptr(buffer) = 8 + len;
    *__dyn_array_len_ptr(buffer) = len;
  } else {
    initial_len = *__dyn_array_len_ptr(buffer);
    buf_begin = __dyn_array_capacity_ptr(buffer);
    uint64_t capacity = *buf_begin;

    if (initial_len + len >= capacity) {
      uint64_t capa = capacity / 2 + capacity + len;
      buf_begin = realloc(buf_begin, size * capa + sizeof(uint64_t) * 2);
      buffer = *arr = buf_begin + 2;
      *__dyn_array_capacity_ptr(buffer) = capa;
    }

    *__dyn_array_len_ptr(buffer) = initial_len + len;
  }

  memcpy(buffer + size * initial_len, from, size * len);
  return initial_len;
}
