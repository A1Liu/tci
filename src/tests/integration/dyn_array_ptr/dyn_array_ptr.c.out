18
15
a b c d e f g h i j k l m n o 
main: abcdefghijklmno
a b c d e f g h i j k l m n o a b c d e f g h i j k l m n o 
